### PR TITLE
[Merged by Bors] - chore: remove autoImplicit from more files

### PR DESCRIPTION
**Pull request by grunweg**
and reduce its scope in a few other instances.
Mostly in `CategoryTheory` and `Data` this time; some `Combinatorics` also.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Module/Torsion.lean [hunk 1]
```diff
@@ -170,7 +170,7 @@ def torsionBySet (s : Set R) : Submodule R M :=
   sInf (torsionBy R M '' s)
 #align submodule.torsion_by_set Submodule.torsionBySet
 
-set_option autoImplicit true in
+universe w in
```
Comment (kmill) [inline Mathlib/Algebra/Module/Torsion.lean:173]: I would suggest using `Type*` in place of `Type w` and delete this line.

File: Mathlib/Algebra/Module/Torsion.lean [hunk 2]
```diff
@@ -185,7 +185,7 @@ def torsion'AddSubMonoid (S : Type w) [CommMonoid S] [DistribMulAction S M] :
     rw [smul_add, mul_smul, mul_comm, mul_smul, hx, hy, smul_zero, smul_zero, add_zero]
   zero_mem' := ⟨1, smul_zero 1⟩
 
-set_option autoImplicit true in
+universe w in
```
Comment (kmill) [inline Mathlib/Algebra/Module/Torsion.lean:188]: Same here.

File: Mathlib/Analysis/RCLike/Basic.lean [hunk 3]
```diff
@@ -51,6 +48,7 @@ local notation "𝓚" => algebraMap ℝ _
 
 open ComplexConjugate
 
+set_option autoImplicit true in
```
Comment (kmill) [inline Mathlib/Analysis/RCLike/Basic.lean:51]: Why is this one necessary?

File: Mathlib/CategoryTheory/ConnectedComponents.lean [hunk 4]
```diff
@@ -48,8 +48,7 @@ def Component (j : ConnectedComponents J) : Type u₁ :=
   FullSubcategory fun k => Quotient.mk'' k = j
 #align category_theory.component CategoryTheory.Component
 
-set_option autoImplicit true in
-instance : Category (Component (j : ConnectedComponents J)) :=
+instance {j : ConnectedComponents J} : Category (Component (j : ConnectedComponents J)) :=
```
Comment (kmill) [inline Mathlib/CategoryTheory/ConnectedComponents.lean:51]:
```suggestion
instance {j : ConnectedComponents J} : Category (Component j) :=
```

File: Mathlib/CategoryTheory/ConnectedComponents.lean [hunk 5]
```diff
@@ -59,12 +58,10 @@ def Component.ι (j : ConnectedComponents J) : Component j ⥤ J :=
   fullSubcategoryInclusion _
 #align category_theory.component.ι CategoryTheory.Component.ι
 
-set_option autoImplicit true in
-instance : Functor.Full (Component.ι (j : ConnectedComponents J)) :=
+instance {j : ConnectedComponents J} : Functor.Full (Component.ι (j : ConnectedComponents J)) :=
```
Comment (kmill) [inline Mathlib/CategoryTheory/ConnectedComponents.lean:61]:
```suggestion
instance {j : ConnectedComponents J} : Functor.Full (Component.ι j) :=
```

File: Mathlib/CategoryTheory/ConnectedComponents.lean [hunk 6]
```diff
@@ -59,12 +58,10 @@ def Component.ι (j : ConnectedComponents J) : Component j ⥤ J :=
...
   FullSubcategory.full _
 
-set_option autoImplicit true in
-instance : Functor.Faithful (Component.ι (j : ConnectedComponents J)) :=
+instance {j : ConnectedComponents J} : Functor.Faithful (Component.ι (j : ConnectedComponents J)) :=
```
Comment (kmill) [inline Mathlib/CategoryTheory/ConnectedComponents.lean:64]:
```suggestion
instance {j : ConnectedComponents J} : Functor.Faithful (Component.ι j) :=
```

File: Mathlib/Data/Multiset/Powerset.lean [hunk 7]
```diff
@@ -140,6 +137,7 @@ theorem revzip_powersetAux' {l : List α} ⦃x⦄ (h : x ∈ revzip (powersetAux
   exact Quot.sound (revzip_sublists' _ _ _ h)
 #align multiset.revzip_powerset_aux' Multiset.revzip_powersetAux'
 
+universe u in
```
Comment (kmill) [inline Mathlib/Data/Multiset/Powerset.lean:140]: It should be possible to change `Type u` below to `Type*` and delete this line. (And delete the porting note.)

**Review by kmill — COMMENTED** [review]
There are just a few comments to address, but everything looks good. Once you fix these, feel free to merge.

bors d+

**Comment by rosborn** [comment]
What is the current guidelines around `variable {α : Type*}` vs `variable {α}`? Is there a preference for one over the other?

File: Mathlib/Data/Nat/Cast/Defs.lean [hunk 8]
```diff
@@ -24,7 +24,7 @@ Preferentially, the homomorphism is written as the coercion `Nat.cast`.
 * `Nat.cast`: Canonical homomorphism `ℕ → R`.
 -/
 
-set_option autoImplicit true
+universe u
```
Comment (rosborn) [inline Mathlib/Data/Nat/Cast/Defs.lean:27]: Could be removed and `Type u` replaced with `Type*`

File: Mathlib/Data/Quot.lean [hunk 9]
```diff
@@ -39,6 +36,7 @@ end Setoid
 namespace Quot
 
 variable {ra : α → α → Prop} {rb : β → β → Prop} {φ : Quot ra → Quot rb → Sort*}
+universe u
```
Comment (rosborn) [inline Mathlib/Data/Quot.lean:39]: This can also be removed. `induction_on` can use `{a : Sort*}` instead.

**Review by rosborn — COMMENTED** [review]
Left a couple comments. I'm assuming we don't want to introduce explicit universes when unnecessary.

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean [hunk 10]
```diff
@@ -914,6 +915,8 @@ structure IsPath {u v : V} (p : G.Walk u v) extends IsTrail p : Prop where
   support_nodup : p.support.Nodup
 #align simple_graph.walk.is_path SimpleGraph.Walk.IsPath
 
+set_option autoImplicit true
```
[the remark this answers is not in the document]
Comment (kmill) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity.lean]: The `set_option autoImplicit true` might be a TODO enough

File: Mathlib/Analysis/RCLike/Basic.lean (hunk 3 above)
Comment (grunweg) [reply Mathlib/Analysis/RCLike/Basic.lean]: It wasn't; thanks for prodding me to remove it!

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean (hunk 10 above)
[the remark this answers is not in the document]
Comment (grunweg) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity.lean]: There's a different reason, I got tired at some point. But I pushed a new commit fixing this; it wasn't so bad after all. :-)

My understanding is that most uses of `autoImplicit` are supposed to be removed; I'm gradually doing so. Feel free to join the effort :-)

File: Mathlib/Data/Array/Defs.lean [hunk 11]
```diff
@@ -11,10 +11,11 @@ This file contains various definitions on `Array`. It does not contain
 proofs about these definitions, those are contained in other files in `Mathlib.Data.Array`.
 -/
 
-set_option autoImplicit true
-
 namespace Array
 
+universe u
+variable {α : Type u}
```
[the remark this answers is not in the document]
Comment (grunweg) [reply Mathlib/Data/Array/Defs.lean:17]: I left this as-is since this file doesn't import the `TypeStar` tactic yet (i.e., `Type*` isn't defined yet).

**Pull request closed without merging**